src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs

```diff
@@ -2,27 +2,164 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Internal;
 using Microsoft.Extensions.Logging;
 using Microsoft.Net.Http.Headers;
 
 namespace Microsoft.AspNetCore.Mvc.Internal
 {
     public class FileResultExecutorBase
     {
+        private PreconditionState _ifMatchState;
+        private PreconditionState _ifNoneMatchState;
+        private PreconditionState _ifModifiedSinceState;
+        private PreconditionState _ifUnmodifiedSinceState;
+
         public FileResultExecutorBase(ILogger logger)
         {
             Logger = logger;
+            _ifMatchState = PreconditionState.Unspecified;
+            _ifNoneMatchState = PreconditionState.Unspecified;
+            _ifModifiedSinceState = PreconditionState.Unspecified;
+            _ifUnmodifiedSinceState = PreconditionState.Unspecified;
+        }
+
+        public enum PreconditionState
+        {
+            Unspecified,
+            NotModified,
+            ShouldProcess,
+            PreconditionFailed,
         }
 
         protected ILogger Logger { get; }
 
+        public PreconditionState GetPreconditionState()
+        {
+            return GetMaxPreconditionState(_ifMatchState, _ifNoneMatchState,
+                _ifModifiedSinceState, _ifUnmodifiedSinceState);
+        }
+
+        private static PreconditionState GetMaxPreconditionState(params PreconditionState[] states)
+        {
+            PreconditionState max = PreconditionState.Unspecified;
+            for (int i = 0; i < states.Length; i++)
+            {
+                if (states[i] > max)
+                {
+                    max = states[i];
+                }
+            }
+            return max;
+        }
+
         protected void SetHeadersAndLog(ActionContext context, FileResult result)
         {
             SetContentType(context, result);
             SetContentDispositionHeader(context, result);
             Logger.FileResultExecuting(result.FileDownloadName);
         }
 
+        private long SetContentLength(ActionContext context, RangeItemHeaderValue range)
+        {
+            long start = range.From.Value;
+            long end = range.To.Value;
+            long length = end - start + 1;
+            var response = context.HttpContext.Response;
+            response.ContentLength = length;
+            return length;
+        }
+
+        protected long SetRangeHeaders(ActionContext context, FileResult result, RangeItemHeaderValue range)
+        {
+            var response = context.HttpContext.Response;
+            response.Headers[HeaderNames.AcceptRanges] = "bytes";
+            var method = context.HttpContext.Request.Method;
+            var isGet = string.Equals("GET", method, StringComparison.OrdinalIgnoreCase);
```

Comment:
Use a constant for get, I believe there is HttpMethodNames somewhere in abstractions